### PR TITLE
PR Review for Ecommerce App

### PR DESCRIPTION
**Pull request by Janaki2399**
Name: Janaki S
Live Link:https://dhrutham-cart.netlify.app/
To Login please use this credential: email - admin@gmail.com, password - admin

I have focussed more on the functionality part. Please review my code and let me know where I can improve.

### CODE REVIEW

File: src/loader-toast-context.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { createContext, useContext, useState } from "react";
+
+const loaderToastContext = createContext();
+
+export function LoaderToastProvider({ children }) {
+  const [toast, setToast] = useState(null);
```
Comment (kushanksriraj) [inline src/loader-toast-context.js]: In useState you have initialized it with null. You are using the state to store string so better initialize it with "", i.e, empty  string.

File: src/loader-toast-context.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { createContext, useContext, useState } from "react";
+
+const loaderToastContext = createContext();
```
Comment (kushanksriraj) [inline src/loader-toast-context.js]: context provider name should be in capital.

File: src/data-context.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        showToast(`Added to ${toastItem}`);
+        hideToast();
+      }
+    } catch (error) {
+      hideToast();
+      alert(error);
+    }
+  }
+
+  async function updateListAndServer({ url, postObject, dispatchType, item }) {
```
Comment (kushanksriraj) [inline src/data-context.js]: How about naming it updateToListAndServer?

File: src/App.js [hunk 4]
```diff
@@ -1,25 +1,54 @@
-import logo from './logo.svg';
+
 import './App.css';
+import { useEffect } from "react";
+import { useDataContext } from "./data-context";
+import { Products } from "./components/Products/Products";
```
Comment (kushanksriraj) [inline src/App.js]: Add an index.js file in components folder and export all components from there, then import them in one line.

File: src/App.js [hunk 5]
```diff
@@ -1,25 +1,54 @@
...
-        <a
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
+      <div class="center-div">
```
Comment (kushanksriraj) [inline src/App.js]: use className instead of class, check for warnings in console.

File: src/reducerFunction.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    case "SHOW_COMPONENT":
+      return {
+        ...state,
+        showComponent: action.payload
+      };
+    default:
+      return state;
+  }
+}
+function checkIfItemExistsInList(list, productItem) {
```
Comment (kushanksriraj) [inline src/reducerFunction.js]: Extract these functions to another file like helper.js and import it.

File: src/components/Cart/QuantityButtons.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+export function QuantityButtons({ cartItem }) {
+  const { updateListAndServer } = useDataContext();
+
+  return (
+    <div>
+      <div class="flex-horizontal margin-top">
+        <button
+          class="icon-btn"
+          style={{ border: "1px solid gray" }}
+          disabled={cartItem.quantity === 1}
```
Comment (kushanksriraj) [inline src/components/Cart/QuantityButtons.js]: check for <=1 then disable the button.

File: src/components/Wishlist/WishListItem.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+export function WishListItem({ wishListItem }) {
+  return (
+    <div
+      className="card card-shadow card-vertical"
+      style={{ maxWidth: "15rem" }}
+    >
+      <div>
+        <img className="card-img" src={wishListItem.image} alt="card-img" />
+      </div>
+      <CardItemContent item={wishListItem} />
```
Comment (kushanksriraj) [inline src/components/Wishlist/WishListItem.js]: Very clean component definition. Nice work.

**Review by kushanksriraj — CHANGES_REQUESTED** [review]

**Comment by saurav49** [comment]
1. after adding to the wishlist, if we click on the heart icon again, it gives an error "request failed with status code 400"
2. total price in the cart component is half hidden.
3. when decreasing the quantity amount for a particular product it is showing an error "request failed with status code 400".

File: src/components/Cart/AddToWishListButton.jsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import { useDataContext } from "../../contexts/data-context";
+
+export function AddToWishListButton({ cartItemId,productId }) {
+  const { removeFromListAndServer, addToListAndServer } = useDataContext();
+
+  function RemoveAndAddItemToListAndServer() {
```
Comment (Xdev200) [inline src/components/Cart/AddToWishListButton.jsx]: logic should be separate from view

**Review by Xdev200 — CHANGES_REQUESTED** [review]
1. For API URL you can create a variable in custom hook and call it everywhere.
2. Update Your README
3. Add Website Address to GitHub

**Comment by Janaki2399** [comment]
> 1. after adding to the wishlist, if we click on the heart icon again, it gives an error "request failed with status code 400"
> 2. total price in the cart component is half hidden.
> 3. when decreasing the quantity amount for a particular product it is showing an error "request failed with status code 400".

Hey there was a typo in the backend url.Please check it now and review

File: src/loader-toast-context.js (hunk 1 above)
Comment (Janaki2399) [reply src/loader-toast-context.js]: Done

File: src/loader-toast-context.js (hunk 2 above)
Comment (Janaki2399) [reply src/loader-toast-context.js]: Done

File: src/App.js (hunk 5 above)
Comment (Janaki2399) [reply src/App.js]: Done

File: src/components/Cart/QuantityButtons.js (hunk 7 above)
Comment (Janaki2399) [reply src/components/Cart/QuantityButtons.js]: Done

**Comment by Janaki2399** [comment]
> 1. For API URL you can create a variable in custom hook and call it everywhere.
> 2. Update Your README
> 3. Add Website Address to Github

Thank you for the suggestions will update the README once I merge it with main branch

File: src/reducerFunction.js (hunk 6 above)
Comment (Janaki2399) [reply src/reducerFunction.js]: Done

File: src/components/Cart/AddToWishListButton.jsx (hunk 9 above)
Comment (Janaki2399) [reply src/components/Cart/AddToWishListButton.jsx]: Done.

File: src/data-context.js (hunk 3 above)
Comment (Janaki2399) [reply src/data-context.js]: We normally say update list right?

**Comment by saurav49** [comment]
The code looks clean, you got most of the view and logic separate.

**Comment by t-rex777** [comment]
everything is very clean and nice, just my personal opinion, you can add padding to navigation menu, it would look better.

**Pull request merged**